Source/JavaScriptCore/runtime/PropertyTable.cpp

```diff
@@ -92,18 +96,34 @@ PropertyTable::PropertyTable(VM& vm, unsigned initialCapacity, const PropertyTab
     : JSCell(vm, vm.propertyTableStructure.get())
     , m_indexSize(sizeForCapacity(initialCapacity))
     , m_indexMask(m_indexSize - 1)
-    , m_index(static_cast<unsigned*>(PropertyTableMalloc::zeroedMalloc(dataSize())))
+    , m_indexVector()
     , m_keyCount(0)
     , m_deletedCount(0)
 {
     ASSERT(isPowerOf2(m_indexSize));
     ASSERT(initialCapacity >= other.m_keyCount);
+    bool isCompact = other.isCompact() && tableCapacity() < UINT8_MAX;
+    m_indexVector = (bitwise_cast<uintptr_t>(PropertyTableMalloc::zeroedMalloc(dataSize(isCompact))) | (isCompact ? isCompactFlag : 0));
+    ASSERT(this->isCompact() == isCompact);
 
-    const_iterator end = other.end();
-    for (const_iterator iter = other.begin(); iter != end; ++iter) {
-        ASSERT(canInsert());
-        reinsert(*iter);
-        iter->key->ref();
+    if (isCompact) {
+        auto* indexVector = bitwise_cast<uint8_t*>(m_indexVector & indexVectorMask);
+        auto* table = tableFromIndexVector(indexVector);
+        other.forEachProperty([&](auto& entry) {
+            ASSERT(canInsert(entry));
+            reinsert(indexVector, table, entry);
+            entry.key()->ref();
+            return IterationStatus::Continue;
+        });
+    } else {
+        auto* indexVector = bitwise_cast<uint32_t*>(m_indexVector & indexVectorMask);
+        auto* table = tableFromIndexVector(indexVector);
+        other.forEachProperty([&](auto& entry) {
+            ASSERT(canInsert(entry));
+            reinsert(indexVector, table, entry);
+            entry.key()->ref();
+            return IterationStatus::Continue;
+        });
```

Comment:
Do we really need the branch here? The two sides are the same, minus how we iterate. But how we iterate is already handled by `forEachProperty`. So I don't think we need this branch with duplicated code. The lambda should just handle it since it's using `auto`

Reply:
Changed it to use `withIndexVector`.